### PR TITLE
Apex callback refactoring

### LINKED ISSUE #4199: Refactor APEX to eliminate circular dependency
The HPX build keeps breaking because CI doesn't test with APEX and because there's a circular dependency between APEX and HPX. This proposed fix will eliminate the dependency, and will be the first step in making APEX a module.
 
The plan:

I think that making the APEX calls using a callback interface is the best
approach.  There are 11 API calls that are currently used, and one object
type.

The object type already has a configurable definition (when APEX isn't
used, it is just defined as a void*).

The complicated bit will be the apex_task_wrapper object, which is a smart
pointer inside the hpx::threads::thread_data object, but I am sure there's
a work around.

I'll start a branch and change the existing APEX calls to callbacks, and
then figure out the required registration interface.

The apex.hpp and apex.cpp files will be replaced with external_timer.hpp and (if necessary) external_timer.cpp, which will define the hpx::util::external_timer class.  The class will provide a callback interface so that APEX can register with HPX at startup.  HPX will make timer calls to `hpx::util::external_timer::start()` instead of `apex::start()`, for example.  The internal HPX smart pointer will change to a generic base class that APEX can extend to implement the task_wrapper object, and an hpx_thread will maintain a `std::shared_ptr<hpx::util::external_timer::task_wrapper>` object instead of `std::shared_ptr<apex::task_wrapper>`.

### PR DESCRIPTION
**Pull request by khuck**
Fixes #4199

## Proposed Changes

  - libhpx no longer contains any APEX symbols.
  - Any libraries or plugins or executables that *do* depend on APEX will still link with it
  - APEX now only has one dependency (or two, if HPXMP is used) - libhpx

## Any background context you want to provide?

This eliminates circular dependencies between APEX and HPX, and is the first step to making APEX a module in HPX.

### CODE REVIEW

File: hpx/util/external_timer.hpp [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+#define HPX_UTIL_WEAK_POST
+#else // not WIN32
+#ifdef __clang__
+#define HPX_UTIL_WEAK_PRE
+#define HPX_UTIL_WEAK_POST __attribute__((weak_import))
+#else // not clang
+#define HPX_UTIL_WEAK_PRE __attribute__((weak))
+#define HPX_UTIL_WEAK_POST
+#endif
+#endif
```
Comment (msimberg) [inline hpx/util/external_timer.hpp]: Where are these needed/used? I don't see any signs of them here or in APEX.

**Comment by msimberg** [comment]
Thanks @khuck! Let me try to summarize your changes to see if I've understood them correctly:

- HPX now exposes an `external_timer` class which allows registering callbacks for various events. HPX doesn't call anything from APEX directly.
- APEX still depends on HPX the usual way. However, it is built as a shared library which initializes the callbacks through a global object in the APEX shared library. Dependent projects need to link to the APEX library, but this is handled through `target_link_libraries(hpx INTERFACE apex)` so that it's transparent to users.

File: examples/1d_stencil/CMakeLists.txt [hunk 2]
```diff
@@ -66,12 +66,13 @@ foreach(example_program ${example_programs})
...
     NOT ("${example_program}" IN_LIST disabled_tests))
 
-    add_hpx_example_test("1d_stencil" ${example_program} ${${example_program}_PARAMETERS})
+  add_hpx_example_test("1d_stencil" ${example_program} ${${example_program}_PARAMETERS})
 
   endif()
 
 endforeach()
 
+
```
[the remark this answers is not in the document]
Comment (khuck) [reply examples/1d_stencil/CMakeLists.txt]: probably not

File: hpx/runtime/threads/thread_data_fwd.hpp [hunk 3]
```diff
@@ -24,11 +24,10 @@
 #include <utility>
 
 #if defined(HPX_HAVE_APEX)
-// forward declare the APEX object
-namespace apex {
+namespace hpx::util::external_timer {
```
[the remark this answers is not in the document]
Comment (khuck) [reply hpx/runtime/threads/thread_data_fwd.hpp]: will do

File: hpx/runtime/threads/thread_data_fwd.hpp [hunk 4]
```diff
@@ -24,11 +24,10 @@
...
     struct task_wrapper;
 }
-using apex_task_wrapper = std::shared_ptr<apex::task_wrapper>;
+using task_wrapper = std::shared_ptr<hpx::util::external_timer::task_wrapper>;
```
[the remark this answers is not in the document]
Comment (khuck) [reply hpx/runtime/threads/thread_data_fwd.hpp]: This should be removed - someone else added that "using" statement and I just updated what it defined.  I can remove it and be explicit instead.

File: hpx/util/external_timer.hpp [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        enable_parent_task_handler_type f);
+
+    namespace external_timer {
+
+    /* HPX provides a smart pointer to a data object that maintains
+     * information about an hpx_thread.  Any library (i.e. APEX)
+     * that wants to use this callback API needs to extend this class.
+     */
+    struct task_wrapper {
+        bool dummy;
```
[the remark this answers is not in the document]
Comment (khuck) [reply hpx/util/external_timer.hpp]: no, I just assumed my struct needed some contents.  I'll try without.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -278,7 +278,12 @@ if (TARGET hpx::amplifier)
 endif()
 
 if (TARGET hpx::apex)
-  target_link_libraries(hpx PRIVATE hpx::apex)
+  # APEX won't get explicitly pulled into libhpx.so any more.  HOWEVER,
+  # we do want to add the APEX link commands to all executables, so 
+  # we use the "INTERFACE" option for target_link_libraries.
+  # Because libhpx_apex is a shared object library, we don't need to specify
+  # the whole archive.
+  target_link_libraries(hpx INTERFACE hpx::apex)
```
[the remark this answers is not in the document]
Comment (khuck) [reply src/CMakeLists.txt]: I am not sure I understand.  You mean the `target_include_directories` on line 291?

File: src/util/external_timer.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    new_task_string_t *new_task_string_function{nullptr};
+    new_task_address_t *new_task_address_function{nullptr};
+    sample_value_t *sample_value_function{nullptr};
+    send_t *send_function{nullptr};
+    recv_t *recv_function{nullptr};
+    HPX_EXPORT update_task_string_t *update_task_string_function{nullptr};
+    HPX_EXPORT update_task_address_t *update_task_address_function{nullptr};
+    HPX_EXPORT start_t *start_function{nullptr};
+    HPX_EXPORT stop_t *stop_function{nullptr};
+    HPX_EXPORT yield_t *yield_function{nullptr};
```
[the remark this answers is not in the document]
Comment (khuck) [reply src/util/external_timer.cpp]: done

File: hpx/util/external_timer.hpp (hunk 1 above)
Comment (khuck) [reply hpx/util/external_timer.hpp]: those are leftover from an aborted solution that wasn't needed (solved with HPX_EXPORT). I will remove them.

**Comment by khuck** [comment]
> Thanks @khuck! Let me try to summarize your changes to see if I've understood them correctly:
> 
> * HPX now exposes an `external_timer` class which allows registering callbacks for various events. HPX doesn't call anything from APEX directly.
> * APEX still depends on HPX the usual way. However, it is built as a shared library which initializes the callbacks through a global object in the APEX shared library. Dependent projects need to link to the APEX library, but this is handled through `target_link_libraries(hpx INTERFACE apex)` so that it's transparent to users.

Yes, that sums it up nicely.  I'm addressing your review comments and will update the branch in a few...

**Comment by khuck** [comment]
@msimberg ok, I addressed your comments and fixed two inspect errors that slipped through.  let me know if I need to address anything else...  the failing tests on (not required) APEX build are lated to apex policies not exiting at shutdown, those have been failing for a while now and aren't related to this PR.  I will have to fix them eventually, of course.

**Comment by khuck** [comment]
I have no idea what Codacy is complaining about - those union members are used in the cpp file

File: src/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply src/CMakeLists.txt]: Sorry, yes, I meant `target_include_directories`...

**Comment by hkaiser** [comment]
@khuck I marked the Codacy 'errors' as resolved an re-triggered the analysis. All should be fine now.

**Comment by khuck** [comment]
Excellent, thanks!

**Pull request merged**